src/js/components/MaskedInput/MaskedInput.js

```diff
@@ -0,0 +1,375 @@
+            valueParts.push({
+              part,
+              beginIndex: valueIndex,
+              endIndex: valueIndex + length - 1,
+            });
+            valueIndex += length;
+            maskIndex += 1;
+            found = true;
+          }
+          length -= 1;
+        }
+        if (!found) {
+          valueIndex = value.length;
+        }
+      } else {
+        const length = Array.isArray(item.length)
+          ? item.length[1]
+          : item.length || value.length - valueIndex;
+        const part = value.slice(valueIndex, valueIndex + length);
+        valueParts.push({
+          part,
+          beginIndex: valueIndex,
+          endIndex: valueIndex + length - 1,
+        });
+        valueIndex += length;
+        maskIndex += 1;
+      }
+    }
+  }
+  return valueParts;
+};
+
+class MaskedInput extends Component {
+  static contextType = ThemeContext;
+
+  static defaultProps = {
+    mask: [],
+  };
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const { mask, value } = nextProps;
+    const { priormask, priorValue } = prevState;
+    if (priormask !== mask || priorValue !== value) {
+      const valueParts = parseValue(mask, value);
+      return { priormask: mask, priorValue: value, valueParts };
+    }
+    return null;
+  }
+
+  state = {};
+
+  inputRef = React.createRef();
+
+  dropRef = React.createRef();
+
+  componentDidUpdate() {
+    const { focused } = this.state;
+    if (focused) {
+      this.locateCaret();
+    }
+  }
+
+  componentWillUnmount() {
+    clearTimeout(this.caretTimeout);
+    clearTimeout(this.blurTimeout);
+  }
+
+  locateCaret = () => {
+    // leave time for caret to be placed after receiving focus
+    clearTimeout(this.caretTimeout);
+    this.caretTimeout = setTimeout(() => {
+      const { mask } = this.props;
+      const { activeMaskIndex, valueParts } = this.state;
+      if (this.inputRef.current) {
+        // determine which mask element the caret is at
+        const caretIndex = this.inputRef.current.selectionStart;
+        let maskIndex;
+        valueParts.some((part, index) => {
+          if (part.beginIndex <= caretIndex && part.endIndex >= caretIndex) {
+            maskIndex = index;
+            return true;
+          }
+          return false;
+        });
+        if (maskIndex === undefined && valueParts.length < mask.length) {
+          maskIndex = valueParts.length; // first unused one
+        }
+        if (maskIndex && mask[maskIndex].fixed) {
+          maskIndex -= 1; // fixed mask parts are never "active"
+        }
+        if (activeMaskIndex !== maskIndex) {
+          // eslint-disable-next-line react/no-did-update-set-state
+          this.setState({ activeMaskIndex: maskIndex, activeOptionIndex: -1 });
+        }
+      }
+    }, 10); // 10ms empirically chosen
+  };
+
+  onFocus = event => {
+    const { onFocus } = this.props;
+    this.locateCaret();
+    this.setState({ focused: true });
```

Comment:
`withFocus` already gives you are `focus` prop. any reason why we are having our own here?

Reply:
Because we need the delay before we set it to false.